cmd/cfgmetagen/meta.go

```diff
@@ -0,0 +1,147 @@
+	"io/ioutil"
+	"path"
+	"reflect"
+	"strings"
+	"time"
+
+	"gopkg.in/yaml.v2"
+)
+
+type fieldMap map[string]*field
+
+type field struct {
+	Type    string      `yaml:",omitempty"`
+	Kind    string      `yaml:",omitempty"`
+	Default interface{} `yaml:",omitempty"`
+	Doc     string      `yaml:",omitempty"`
+	Meta    string      `yaml:",omitempty"`
+	Fields  fieldMap    `yaml:",omitempty"`
+}
+
+// genMeta creates a `cfgMeta.yaml` file in the directory of the passed-in
+// config instance. The yaml file contains the recursive field names, types,
+// tags, comments, and default values for the config struct.
+func genMeta(cfg interface{}) {
+	cfgVal := reflect.ValueOf(cfg)
+	cfgType := cfgVal.Type()
+	field := &field{
+		Type:   cfgType.String(),
+		Fields: fieldMap{},
+	}
+	buildField(cfgVal, field)
+	marshaled, err := yaml.Marshal(field)
+	if err != nil {
+		panic(err)
+	}
+	pth := typePath(cfgType.Elem())
+	err = ioutil.WriteFile(path.Join(pth, "cfgMeta.yaml"), marshaled, 0600)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func buildField(v reflect.Value, f *field) {
+	if v.Kind() == reflect.Ptr {
+		buildField(v.Elem(), f)
+	}
+	if v.Kind() != reflect.Struct {
+		return
+	}
+	comments := fieldCommentsForStruct(v)
+	for i := 0; i < v.NumField(); i++ {
+		structField := v.Type().Field(i)
+		yamlName, meta := splitTag(structField.Tag)
+		if yamlName == "-" {
+			continue
+		}
+		fv := v.Field(i)
+		next := f
+		if meta != "squash" {
+			doc := comments[structField.Name]
+			name := yamlName
+			if name == "" {
+				name = strings.ToLower(structField.Name)
+			}
+			kindStr := fv.Kind().String()
+			typeStr := fv.Type().String()
+			if typeStr == kindStr {
+				typeStr = "" // omit if redundant
+			}
+			next = &field{
+				Type:   typeStr,
+				Kind:   kindStr,
+				Doc:    doc,
+				Meta:   meta,
+				Fields: fieldMap{},
+			}
+			f.Fields[name] = next
+		}
+		handleKinds(fv, next)
+	}
+}
+
+func handleKinds(v reflect.Value, f *field) {
+	switch v.Kind() {
+	case reflect.Struct:
+		buildField(v, f)
+	case reflect.Ptr:
+		if v.IsNil() {
+			buildField(reflect.New(v.Type().Elem()), f)
+		} else {
+			buildField(v.Addr(), f)
+		}
+	case reflect.Slice:
+		e := v.Type().Elem()
+		if e.Kind() == reflect.Struct {
+			buildField(reflect.New(e), f)
+		} else if e.Kind() == reflect.Ptr {
+			buildField(reflect.New(e.Elem()), f)
+		}
+	case reflect.String:
+		if v.String() != "" {
+			f.Default = v.String()
+		}
+	case reflect.Bool:
+		f.Default = v.Bool()
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		if v.Int() != 0 {
+			if v.Type() == reflect.TypeOf(time.Duration(0)) {
+				f.Default = time.Duration(v.Int()).String()
+			} else {
+				f.Default = v.Int()
+			}
+		}
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		if v.Uint() != 0 {
+			f.Default = v.Uint()
+		}
+	}
+}
+
+func splitTag(s reflect.StructTag) (string, string) {
+	ms := strings.TrimPrefix(string(s), "mapstructure:")
```

Comment:
Will need to fully parse (ie something like https://github.com/fatih/structtag) the tags since there's no guarantee mapstructure is the first tag.

Reply:
Done.